runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1031,64 +1379,81 @@ static iree_status_t iree_hal_hip_device_queue_alloca(
+      (device->supports_memory_pools ||
+       iree_hal_hip_allocator_isa(iree_hal_device_allocator(base_device)))) {
     iree_hal_buffer_t* buffer = NULL;
 
-    IREE_RETURN_IF_ERROR(iree_hal_hip_memory_pools_prepare_buffer(
-        &device->memory_pools, device->hip_dispatch_stream, pool, params,
-        allocation_size, &buffer));
+    status = iree_hal_hip_device_pepare_async_alloc(device, params,
+                                                    allocation_size, &buffer);
 
-    iree_status_t status = iree_hal_deferred_work_queue_enqueue_alloc(
-        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
+    iree_hal_hip_device_semaphore_buffer_operation_callback_data_t*
+        callback_data = NULL;
     if (iree_status_is_ok(status)) {
-      status = iree_hal_deferred_work_queue_issue(device->work_queue);
+      status = iree_hal_hip_device_make_buffer_callback_data(
+          device, device->host_allocator, queue_affinity, wait_semaphore_list,
+          signal_semaphore_list, buffer,
+          IREE_HAL_HIP_DEVICE_SEMAPHORE_OPERATION_ASYNC_ALLOC, &callback_data);
     }
-    if (iree_status_is_ok(status)) {
+    if (iree_status_is_ok(status) && wait_semaphore_list.count == 0) {
+      status = iree_hal_hip_dispatch_thread_add_dispatch(
+          device->topology.devices[device_ordinal].dispatch_thread,
+          &iree_hal_hip_device_perform_buffer_operation_now, callback_data);
       *out_buffer = buffer;
-    } else {
-      iree_hal_hip_buffer_set_allocation_empty(buffer);
-      iree_hal_resource_release(&buffer->resource);
+      IREE_TRACE_ZONE_END(z0);
+      return status;
     }
-    return status;
-  } else if (!iree_all_bits_set(params.type,
-                                IREE_HAL_MEMORY_TYPE_HOST_VISIBLE) &&
-             iree_hal_hip_allocator_isa(
-                 iree_hal_device_allocator(base_device))) {
-    iree_hal_buffer_t* buffer = NULL;
-
-    IREE_RETURN_IF_ERROR(iree_hal_hip_device_pepare_async_alloc(
-        device, params, allocation_size, &buffer));
 
-    iree_status_t status = iree_hal_deferred_work_queue_enqueue_alloc(
-        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
     if (iree_status_is_ok(status)) {
-      status = iree_hal_deferred_work_queue_issue(device->work_queue);
+      for (iree_host_size_t i = 0;
+           i < wait_semaphore_list.count && iree_status_is_ok(status); ++i) {
+        iree_hal_resource_retain(wait_semaphore_list.semaphores[i]);
+        status = iree_status_join(
+            status,
+            iree_hal_hip_semaphore_notify_work(
+                wait_semaphore_list.semaphores[i],
+                wait_semaphore_list.payload_values[i],
+                device->topology.devices[device_ordinal].device_event_pool,
+                &iree_hal_hip_device_semaphore_buffer_operation_callback,
+                callback_data));
+      }
+    } else {
+      iree_allocator_free(device->host_allocator, callback_data);
```

Comment:
I suspect freeing is invalid - the callback data owns things - you need proper cleanup code for each callback data. similar to another file, consider each bit of duplicate cleanup logic as being a source of bugs and that there has to be a very strong justification to have it - outside of very rare situations (that are still often better with more structure, like pooling) there should only be one way to free/deinitialize an object. if this were C++ it'd be like having the logic to cleanup an object in its destructor - you can't have multiple destructors and you want to keep destruction logic together with the owner - here we just call our destructors names like free/destroy/deinitialize instead of ~.